pkg/controller/attach_detach.go

```diff
@@ -573,154 +559,162 @@ func reportError(azVolumeAttachment *azdiskv1beta2.AzVolumeAttachment, state azd
+			hasSynced = true
+			return volumeHandle, volumeContext, vInfo, hasSynced, nil
+		}
+		volumeHandle = vInfo.handle
+		volumeContext = vInfo.attributes
+	} else {
+		w.Logger().Infof("getting volumeHandle and volumeContext from PersistentVolume for volume (%s)", volumeName)
+		volumeHandle, volumeContext, err = r.getVolumeInfoFromPV(ctx, volumeName)
+		if err != nil {
+			w.Logger().Errorf(err, "failed to find volume info from PersistentVolume for volume (%s)", volumeName)
+			return "", nil, vInfo, hasSynced, err
+		}
+		vInfo = generateVolumeInfo(volumeHandle, volumeContext)
+		r.volumeInfos.Store(volumeName, vInfo)
+	}
+	return volumeHandle, volumeContext, vInfo, hasSynced, nil
+}
 
-				azVolumeAttachment, err = r.azClient.DiskV1beta2().AzVolumeAttachments(r.config.ObjectNamespace).Update(ctx, azVolumeAttachment, metav1.UpdateOptions{})
-				if err != nil {
-					w.Logger().Errorf(err, "failed to update AzVolumeAttachment (%s) for volume (%s) and node (%s): %v", azVolumeAttachmentName, *pvName, nodeName, err)
-					return syncedVolumeAttachments, volumesToSync, err
-				}
-			} else {
-				statusUpdateRequired = false
-			}
+func (r *ReconcileAttachDetach) recreateAzVolumeAttachmentFromVolumeInfo(ctx context.Context, volumeAttachmentName, azVolumeAttachmentName, volumeName, nodeName string) error {
+	w, _ := workflow.GetWorkflowFromContext(ctx)
+	volumeHandle, volumeContext, volumeInfo, hasSynced, err := r.loadOrStoreVolumeInfo(ctx, volumeName, azVolumeAttachmentName)
+	if hasSynced {
+		// this azVolumeAttachment has already been recreated
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+	attached, err := r.isAttached(ctx, volumeHandle, nodeName)
+	if !attached && err == nil {
+		// disk is not attached to node. no need to recreate AzVolumeAttachment.
+		return nil
+	}
+	if err != nil {
+		// when it failed to check if the disk is attached, we can still recreate the AzVolumeAttachment based on the previous information we got.
+		// so we will not miss any AzVolumeAttachment
+		w.Logger().Errorf(err, "failed to check if disk (%s) is attached to node (%s).", volumeHandle, nodeName)
+	}
+	role := azdiskv1beta2.ReplicaRole
+	if volumeAttachmentName != "" {
+		r.azVolumeAttachmentToVaMap.Store(azVolumeAttachmentName, volumeAttachmentName)
+		// if it has volumeAttachmentName, means it's a Primary AzVolumeAttachment
+		role = azdiskv1beta2.PrimaryRole
+	}
+	desiredAzVolumeAttachment := r.generateDesiredAzVolumeAttachment(azVolumeAttachmentName, nodeName, volumeName, volumeHandle, role, volumeContext)
+	err = r.recreateAzVolumeAttachment(ctx, azVolumeAttachmentName, desiredAzVolumeAttachment)
+	if err != nil {
+		w.Logger().Errorf(err, "failed to recreate AzVolumeAttachment (%s)", azVolumeAttachmentName)
+		return err
+	}
+	volumeInfo.syncedAzVolumeAttachments[azVolumeAttachmentName] = true
+	r.volumeInfos.Store(volumeName, volumeInfo)
```

Comment:
If you store the `volumeInfo` in the map by pointer rather than by value, you wouldn't have to do an update here.